File: Ruby_on_Rails/Avanzado/ecommerce/app/models/producto.rb
class Producto < ApplicationRecord
  belongs_to :categoria
  has_many :items_pedido
  has_many :pedidos, through: :items_pedido
  has_one_attached :imagen

  validates :nombre, presence: true
  validates :precio, numericality: { greater_than: 0 }
  validates :stock, numericality: { greater_than_or_equal_to: 0 }

  scope :disponibles, -> { where('stock > 0') }
  scope :por_categoria, ->(cat) { joins(:categoria).where(categorias: { nombre: cat }) }

  def disponible?
    stock > 0
  end

  def reducir_stock!(cantidad)
    raise 'Stock insuficiente' if stock < cantidad
    decrement!(:stock, cantidad)
  end
end

File: Laravel/Avanzado/ecommerce/app/Models/Producto.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class Producto extends Model
{
    use SoftDeletes;

    protected $fillable = ['nombre', 'descripcion', 'precio', 'stock', 'categoria_id'];

    protected $casts = [
        'precio' => 'decimal:2',
    ];

    public function categoria()
    {
        return $this->belongsTo(Categoria::class);
    }

    public function itemsPedido()
    {
        return $this->hasMany(ItemPedido::class);
    }

    public function scopeDisponibles($query)
    {
        return $query->where('stock', '>', 0);
    }

    public function reducirStock(int $cantidad): void
    {
        if ($this->stock < $cantidad) {
            throw new \RuntimeException("Stock insuficiente para {$this->nombre}");
        }
        $this->decrement('stock', $cantidad);
    }
}
